### PR TITLE
Progressive Font loading

### LINKED ISSUE #652: Investigate progressive font loading
Load the qpc hafs font until the v1/v2 fonts load. Let's chat more about it.

E.g. website that does this is the reading mode on Surah 1: al-Fatihah – QuranOnline.net

### PR DESCRIPTION
**Pull request by osamasayed**
### Summary
This PR adds a fallback font for QCF's font V1 and V2 until the font of each page finishes downloading.
Closes #652

### CODE REVIEW

File: src/components/QuranReader/QuranReaderBody.tsx [hunk 1]
```diff
@@ -1,9 +1,10 @@
+/* eslint-disable i18next/no-literal-string */
```
Comment (muhajirdev) [inline src/components/QuranReader/QuranReaderBody.tsx:1]: nit: does the `useQcfFont(quranReaderStyles.quranFont, verses)` contains literal string?

Comment (osamasayed) [reply src/components/QuranReader/QuranReaderBody.tsx:1]: No. It used to thanks for catching this :)

**Comment by muhajirdev** [comment]
I think functionality wise, aside from the font-size issue on first load. looks good.
not sure how hard to fix the font-size issue.

File: src/components/Verse/VerseText.tsx [hunk 2]
```diff
@@ -26,12 +28,19 @@ type VerseTextProps = {
...
     selectQuranReaderStyles,
     shallowEqual,
   ) as QuranReaderStyles;
   const [firstWord] = words;
+  let isFontLoaded = true;
+  if (isQCFFont(quranFont)) {
+    isFontLoaded = loadedFonts.includes(
+      `p${firstWord.pageNumber}-${quranFont.replace('code_', '')}`,
+    );
+  }
```
Comment (tifa2UP) [inline src/components/Verse/VerseText.tsx:43]: should be wrapped inside of a useMemo

File: src/components/dls/QuranWord/GlyphWord.module.scss [hunk 3]
```diff
@@ -1,3 +1,7 @@
 .styledWord {
   line-height: normal;
 }
+
+.uthmanicText {
```
Comment (tifa2UP) [inline src/components/dls/QuranWord/GlyphWord.module.scss:5]: .fallbackText?

File: src/components/dls/QuranWord/GlyphWord.tsx [hunk 4]
```diff
@@ -1,19 +1,45 @@
...
 import styles from './GlyphWord.module.scss';
 
+import { QuranFont } from 'types/QuranReader';
+import Word from 'types/Word';
+
 type UthmaniWordTextProps = {
-  text: string;
-  pageNumber: number;
-  font: string;
+  word: Word;
```
Comment (tifa2UP) [inline src/components/dls/QuranWord/GlyphWord.tsx:12]: don't pass the full word object, only the values that you need

File: src/utils/fontFaceHelper.ts [hunk 5]
```diff
@@ -7,39 +7,25 @@ const QCFFontCodes = [QuranFont.MadaniV1, QuranFont.MadaniV2];
...
 
-const buildV1FontFace = (pageNumber: string | number) => `@font-face {
-        font-family: 'p${pageNumber}-v1';
-        src: local(QCF_P${String(pageNumber).padStart(3, '0')}),
-        url('/fonts/v1/woff2/p${pageNumber}.woff2') format('woff2'),
-        url('/fonts/v1/woff/p${pageNumber}.woff') format('woff'),
-        url('/fonts/v1/ttf/p${pageNumber}.ttf') format('truetype');
-        font-display: swap;
-    }`;
+export const getFontFaceSource = (isV1: boolean, pageNumber: number): string =>
```
Comment (tifa2UP) [inline src/utils/fontFaceHelper.ts:25]: Change the name to be more specific to v1/v2. We should also add comments

File: src/utils/fontFaceHelper.ts [hunk 6]
```diff
@@ -7,39 +7,25 @@ const QCFFontCodes = [QuranFont.MadaniV1, QuranFont.MadaniV2];
 
 export const isQCFFont = (font: QuranFont) => QCFFontCodes.includes(font);
 
-const getPages = (verses: Verse[]) => {
+/**
+ * Get the page numbers of a group of verses but getting
```
Comment (tifa2UP) [inline src/utils/fontFaceHelper.ts:11]: by

File: src/utils/fontFaceHelper.ts [hunk 7]
```diff
@@ -7,39 +7,25 @@ const QCFFontCodes = [QuranFont.MadaniV1, QuranFont.MadaniV2];
...
+ * the page number of the first verse, the page number of the last verse
+ * and generating the range between both of them.
+ *
+ * @param {Verse[]} verses
+ * @returns {number[]}
+ */
+export const getVersesPages = (verses: Verse[]): number[] => {
```
Comment (tifa2UP) [inline src/utils/fontFaceHelper.ts:18]: getPagesByVerses

File: src/styles/_utility.scss [hunk 8]
```diff
@@ -24,6 +24,7 @@
   );
 }
 
+// NOTE: IF WE CHANGE ANY VALUE HERE, WE SHOULD ALSO ADAPT THE NEW VALUE TO THE FALLBACK TEXT INSIDE GLYPHWORD.MODULE.SCSS
```
Comment (tifa2UP) [inline src/styles/_utility.scss:27]: love the caps. Missed `IMPORTANT NOTE:`

File: src/components/dls/QuranWord/GlyphWord.module.scss [hunk 9]
```diff
@@ -1,3 +1,67 @@
+@use "src/styles/breakpoints";
+
 .styledWord {
   line-height: normal;
 }
+
+.fallbackText {
```
Comment (tifa2UP) [inline src/components/dls/QuranWord/GlyphWord.module.scss:7]: thank you! Can you add a comment here explaining why we do this?

**Review by tifa2UP — APPROVED** [review]

File: src/styles/_utility.scss (hunk 8 above)
Comment (osamasayed) [reply src/styles/_utility.scss:27]: haha added :)

**Pull request merged**